bgpd/bgpd.c

```diff
@@ -278,17 +279,31 @@ void bgp_router_id_zebra_bump(vrf_id_t vrf_id, const struct prefix *router_id)
 			if (bgp->inst_type == BGP_INSTANCE_TYPE_VRF)
 				continue;
 
-			bgp->router_id_zebra = router_id->u.prefix4;
-			if (!bgp->router_id_static.s_addr)
-				bgp_router_id_set(bgp, &router_id->u.prefix4);
+			bgp->router_id_zebra = *prefix;
+			if (!bgp->router_id_static.s_addr) {
+				if(bgp_router_id_valid(bgp) < 0) {
```

Comment:
As i understand, Router ID update from Zebra is not handled if existing router_id is present in BGP Address hash (probably as an interface address).

Let us take a configuration in this sequence:-
------------------------------
router-id 1.1.1.1
Int lo
  ip address 1.1.1.1/32

router-id 2.2.2.2 ----> Zebra updates its router-id and intimates BGP. BGP ignores this as 1.1.1.1 exist in address hash

int lo
  no ip address 1.1.1.1/32 ---> Zebra deletes this from IFC and sends address delete to BGP. BGP does not update its router id still. 
------------------------------
Looks like the handling is not consistent with your fix as the necessary and sufficient condition you had checked is no longer true. DO we need to update the router-id in this case in BGP ?